rpc/connection_log.go

```diff
@@ -0,0 +1,155 @@
+// implementation that does structural logging may ignore `format` completely
+// if `msgKey` is provided in logField.
+type connectionLog interface {
+	Warning(format string, fields ...logField)
+	Debug(format string, fields ...logField)
+	Info(format string, fields ...logField)
+}
+
+type connectionLogUnstructured struct {
+	LogOutput
+	logPrefix string
+}
+
+func newConnectionLogUnstructured(
+	logOutput LogOutput, prefix string) *connectionLogUnstructured {
+	randBytes := make([]byte, 4)
+	rand.Read(randBytes)
+	return &connectionLogUnstructured{
+		LogOutput: logOutput,
+		logPrefix: strings.Join(
+			[]string{prefix, hex.EncodeToString(randBytes)}, " "),
+	}
+}
+
+func formatLogFields(f string, lf ...logField) string {
+	fields := make([]interface{}, 0, len(lf))
+	for _, lf := range lf {
+		fields = append(fields, lf)
+	}
+	return fmt.Sprintf(f, fields...)
+}
+
+func (l *connectionLogUnstructured) Warning(
+	format string, fields ...logField) {
+	l.LogOutput.Warning("(%s) %s", l.logPrefix,
+		formatLogFields(format, fields...))
+}
+
+func (l *connectionLogUnstructured) Debug(
+	format string, fields ...logField) {
+	l.LogOutput.Debug("(%s) %s", l.logPrefix,
+		formatLogFields(format, fields...))
+}
+
+func (l *connectionLogUnstructured) Info(
+	format string, fields ...logField) {
+	l.LogOutput.Info("(%s) %s", l.logPrefix,
+		formatLogFields(format, fields...))
+}
+
+// LogrusEntry and LogrusLogger define methods we need from logrus to avoid
+// pulling logrus as dependency.
+type LogrusEntry interface {
+	Debug(args ...interface{})
+	Info(args ...interface{})
+	Warning(args ...interface{})
+}
+
+// LogrusLogger maps to *logrus.Logger, but will need an adapter that converts
+// map[string]interface{} to logrus.Fields, and adapt logrus.Entry to
+// LogrusEntry.
+type LogrusLogger interface {
+	WithFields(map[string]interface{}) LogrusEntry
+}
+
+type connectionLogLogrus struct {
+	log LogrusLogger
+
+	section   string
+	randBytes string
+}
+
+func newConnectionLogLogrus(
+	log LogrusLogger, section string) *connectionLogLogrus {
+	randBytes := make([]byte, 4)
+	rand.Read(randBytes)
+	return &connectionLogLogrus{
+		log:       log,
+		section:   section,
+		randBytes: hex.EncodeToString(randBytes),
+	}
+}
+
+func (l *connectionLogLogrus) getLogEntryAndExtractMsg(
+	format string, fields []logField) (entry LogrusEntry, msg string) {
+	_, file, line, ok := runtime.Caller(2)
+	if !ok {
+		file, line = "unknown", -1
+	}
+	mFields := make(map[string]interface{})
+	for _, f := range fields {
+		mFields[f.key] = f.value
+	}
+	mFields["section"] = l.section
+	mFields["identifier"] = l.randBytes
+	mFields["file"], mFields["line"] = file, line
+
+	if msgI, ok := mFields[msgKey]; ok {
+		if msg, ok = msgI.(string); !ok {
+			msg = fmt.Sprintf("%s", msgI)
+		}
+	} else {
+		// If "msg" isn't present in fields, just use a formatted string. It
+		// won't index well but it's better than losing information.
+		msg = formatLogFields(format, fields...)
```

Comment:
So we're duplicating the info here? Why? We'll be logging them anyway as part of the fields. I'd recommend passing a blank message instead.

Reply:
Not duplicating; this is in the logrus logger, which usually ignores `format` completely since it doesn't get indexed well. This is in case `msg` is missing, and we'd still have something in that field so we know what it is in kibana.

Reply:
But I don't think we need anything in that field. It won't be useful, since the remainder of the fields will have the valid metadata. In fact, I'd argue that "" is a useful value for "msg" in that case, since then we can search for log lines that didn't have an explicit format set. Whereas arbitrarily formatting all the log fields won't be useful at all.

Reply:
That's a good point; thanks!. Actually would you be OK with a `"<empty>"`? I'm not sure if `""` gets indexed well in kibana.

Reply:
Why would it be any different?

Reply:
`""` is empty and for example `if (!"") { console.log("this prints") }` in javascript. I know kibana backend is not javascript but was just concerned about it.

Reply:
So your concern is because there's no way to communicate "not present" to logrus for the `msg` field? If that's the case, I'm fine with "<empty>".

Reply:
Yup; sorry this wasn't clear. Will do!